src/include/daos_srv/ras.h

```diff
@@ -81,29 +57,155 @@ ras_event_sev_enum_to_name(enum ras_event_sev severity)
 	case RAS_SEV_ERROR:
 		return "ERROR";
 	case RAS_SEV_INFO:
+	default:
 		return "INFO";
 	}
-
-	return RAS_SEV_UNKNOWN_STR;
 }
 
 enum ras_event_type {
 	/* ANY is a special case to match all types */
 	RAS_TYPE_ANY	= 0,
-	RAS_TYPE_RANK_STATE_CHANGE,
-	RAS_TYPE_INFO_ONLY,
+	RAS_TYPE_STATE_CHANGE,
+	RAS_TYPE_INFO,
 };
 
 static inline char *
-ras_event_type_enum_to_name(enum ras_event_type type)
+ras_event_type2str(enum ras_event_type type)
 {
 	switch (type) {
-	case RAS_TYPE_RANK_STATE_CHANGE:
-		return "RANK_STATE_CHANGE";
-	case RAS_TYPE_INFO_ONLY:
-		return "INFO_ONLY";
+	case RAS_TYPE_STATE_CHANGE:
+		return "STATE_CHANGE";
+	case RAS_TYPE_INFO:
+	default:
+		return "INFO";
 	}
+}
+
+/**
+ * RAS event should be both printed to debug log and sent to the control
+ * plane.
+ * XXX: only print to stdout for now
+ */
+static inline void
+d_ras_raise(const char *id, enum ras_event_type type, enum ras_event_sev sev,
+	    char *hid, d_rank_t rank, char *jid, uuid_t puuid, uuid_t cuuid,
+	    daos_obj_id_t *oid, const char *cop, const char *msg,
+	    const char *data, ...)
+{
+	struct timeval	tv;
+	struct tm	*tm;
+	struct utsname	uts;
+	va_list		ap;
+
+	printf("&&& RAS EVENT id: [%s]", id);
+
+	/** print timestamp */
+	(void) gettimeofday(&tv, 0);
+	tm = localtime(&tv.tv_sec);
+	if (tm) {
+		printf(" ts: [%04d/%02d/%02d-%02d:%02d:%02d.%02ld]",
+		       tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
+		       tm->tm_hour, tm->tm_min, tm->tm_sec,
+		       (long int) tv.tv_usec / 10000);
+	}
+
+	/** XXX: nodename should be fetched only once and not on every call */
+	(void) uname(&uts);
+	printf(" host: [%s]", uts.nodename);
+
+	printf(" type: [%s] sev: [%s]", ras_event_type2str(type),
+	       ras_event_sev2str(sev));
+
+	if (hid)
+		printf(" hwid: [%s]", hid);
+
+	if (rank)
+		printf(" rank: [%u]", rank);
 
-	return RAS_TYPE_UNKNOWN_STR;
+	if (jid)
+		printf(" jobid: [%s]", jid);
+
+	if (!uuid_is_null(puuid))
+		printf(" puuid: ["DF_UUIDF"]", DP_UUID(puuid));
+
+	if (!uuid_is_null(cuuid))
+		printf(" cuuid: ["DF_UUIDF"]", DP_UUID(cuuid));
+
+	if (oid)
+		printf(" oid: ["DF_OID"]", DP_OID(oid));
+
+	if (cop)
+		printf(" control_op: [%s]", cop);
+
+	/** print msg */
+	printf(" msg: [%s]", msg);
+
+	/** print data blob */
+	printf(" data: [");
```

Comment:
In DAOS-6063 I started off passing opaque data blobs from event producer all the way to consumer and encoding/decoding at the edges, after a request from @mjmac I changed this approach to provide specific payload types in protobuf definitions, the idea being that the event specific structures would be available either end of the dRPC channel (for use in control and data planes, Go & C). When raising event in iosrv, we could encode blob and execute dRPC which would in turn decode into a specific payload type to be serialised but this seems inefficient. I suggest we either stick to transporting blob including when serialising/deserialising protobuf messages (bytes type) or alternatively populate the protobuf generated specific payload types when raising the event (which will log and forward to MS over dRPC). we could go for a hybrid approach where we have an "opaque" payload type and when raising events, either a proto defined payload or opaque blob is chosen (former would normally be when event needs to be acted upon within the control plane, the latter would be reserved for events that just need to be forwarded to a consumer outside the control plane).

Reply:
When accepting variable number of const char *data parameters, these would have to be converted to e.g. rank lists in the case of a pool service replica list update event, alternatively these could be transferred over the wire in proto message format as repeated string parameter to be decoded when necessary with the producer and consumer of the event having implicit knowledge of the event specific payload schema.